protocols/jvss/handlers.go

```diff
@@ -0,0 +1,173 @@
+package jvss
+
+import (
+	"fmt"
+
+	"github.com/dedis/cothority/lib/dbg"
+	"github.com/dedis/cothority/lib/sda"
+	"github.com/dedis/crypto/poly"
+)
+
+// SecInitMsg are used to initialise new shared secrets both long- and
+// short-term.
+type SecInitMsg struct {
+	Src  int
+	SID  string
+	Deal []byte
+}
+
+// SecConfMsg are used to confirm to other peers that we have finished setting
+// up the shared secret.
+type SecConfMsg struct {
+	Src int
+	SID string
+}
+
+// SigReqMsg are used to send signing requests.
+type SigReqMsg struct {
+	Src int
+	SID string
+	Msg []byte
+}
+
+// SigRespMsg are used to reply to signing requests.
+type SigRespMsg struct {
+	Src int
+	SID string
+	Sig *poly.SchnorrPartialSig
+}
+
+// WSecInitMsg is a SDA-wrapper around SecInitMsg.
+type WSecInitMsg struct {
+	*sda.TreeNode
+	SecInitMsg
+}
+
+// WSecConfMsg is a SDA-wrapper around SecConfMsg.
+type WSecConfMsg struct {
+	*sda.TreeNode
+	SecConfMsg
+}
+
+// WSigReqMsg is a SDA-wrapper around SigReqMsg.
+type WSigReqMsg struct {
+	*sda.TreeNode
+	SigReqMsg
+}
+
+// WSigRespMsg is a SDA-wrapper around SigRespMsg.
+type WSigRespMsg struct {
+	*sda.TreeNode
+	SigRespMsg
+}
+
+func (jv *JVSS) handleSecInit(m WSecInitMsg) error {
+	msg := m.SecInitMsg
+
+	// Initialise shared secret
+	if err := jv.initSecret(msg.SID); err != nil {
+		return err
+	}
+
+	// Unmarshal received deal
+	deal := new(poly.Deal).UnmarshalInit(jv.info.T, jv.info.R, jv.info.N, jv.keyPair.Suite)
+	if err := deal.UnmarshalBinary(msg.Deal); err != nil {
+		return err
+	}
+
+	// Buffer received deal for later
+	secret, ok := jv.secrets[msg.SID]
+	if !ok {
+		return fmt.Errorf("Error, shared secret does not exist")
+	}
+	secret.deals[msg.Src] = deal
+
+	// Finalise shared secret
+	if err := jv.finaliseSecret(msg.SID); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (jv *JVSS) handleSecConf(m WSecConfMsg) error {
+	msg := m.SecConfMsg
+
+	secret, ok := jv.secrets[msg.SID]
+	if !ok {
+		return fmt.Errorf("Error, shared secret does not exist")
+	}
+
+	secret.mtx.Lock()
+	secret.numConfs++
+	secret.mtx.Unlock()
+
+	dbg.Lvl2(fmt.Sprintf("Node %d: %s confirmations %d/%d", jv.nodeIdx(), msg.SID, secret.numConfs, len(jv.nodeList)))
+
+	// Check if we have enough confirmations to proceed
+	if (secret.numConfs == len(jv.nodeList)) && (msg.SID == LTSS || msg.SID == fmt.Sprintf("%s%d", STSS, jv.nodeIdx())) {
```

Comment:
Btw. I find this construction a bit awkward, especially the second part where I check which SID the message has in order to be able to notify the correct node that all confirmations (for establishing the shared secret) have arrived.

If you have a better idea how to solve that, I am open for suggestions. :)